### ISSUE TITLE
shielded resource logic check

### ISSUE THREAD

**Issue by XuyangSong** [anoma/green#390]
customer: @mariari 
performer: @XuyangSong 
deadline: null
estimated: null
started: null
actual: null
completed: null
confirmed: null
dependencies:
- #644 
- #495 

---

<put issue details here>

The resource label can be extracted from a transparent resource simply by compressing the resource logic program since the logic is open. Additionally, we can extract the resource label from a shielded resource in SNARK (R1CS or Plonkish) by compressing only the necessary parts of the verifying key.

The STARK systems (based on AIR) don't have explicit verifying keys. We need to dig into the CairoAIR structure and extract the "logic" parts where we exclude instance data and make sure different resource logics lead to different labels.

According to the Note in stone-prover:
> Note: The verifier only checks that the proof is consistent with the public input section that appears in the proof file. The public input section itself is not checked. For example, the verifier does not check what CairoZero program is being proved, or that the builtins memory segments are of valid size. These things need to be checked externally.

The bottom line is, we need extract the info about what constraints the verifier is checking from CairoAIR and wrap it up in the resource label. The fact that the logic and instance data are all mixed up in Cairo's public memory can be annoying, so I'll dig into it a bit more.
